Fix #21215: Wrong location for "not a member" error when constructing a struct literal

Ay1nDas opened the pull request:
Fixes #21215 
Fixed Error msg from:
``` test.d(6): Error: `y` is not a member of `S` ```
To:
``` test.d(5): Error: `y` is not a member of `S` ```
Now it points to the correct location.

Ay1nDas committed:
Added fix

Ay1nDas committed:
Changing passing argLabels by refrence to by pointer.

Ay1nDas committed:
Fixing Error msg for not a member in struct

thewilsonator commented on compiler/src/dmd/expression.d:
```diff
@@ -3288,12 +3288,19 @@ struct ArgumentList
 {
     Expressions* arguments; // function arguments
     Identifiers* names;     // named argument identifiers
+    ArgumentLabel[] argLabels; // named argument labels
```
you will need yo update the corresponding C++ header with a `DArray<ArgumentLabel>`, also `ArgumentLabel` will need to be `extern(C++)`. You will also need to update `frontend.h` accordingly.

thewilsonator commented on compiler/test/fail_compilation/test21215.d:
```diff
@@ -0,0 +1,45 @@
+    
+    2
+  );
+}
```
add an extra newline here

Ay1nDas replied:
Thank you for the suggestions.

Ay1nDas committed:
Updated C++ header and added compatibility b/w names and argLabels

Ay1nDas committed:
Changed ArgumentLabel to Array!() and Updated the relevent C++ headers

Ay1nDas committed:
Fixing errors

Ay1nDas committed:
Added ArgumentLabels to astbase.d

Ay1nDas commented:
The `frontend.d` was generated on my PC using `rdmd run.d cxx-headers-test BUILD=debug AUTO_UPDATE=1`
Then why is the generated one in Circleci different?
Both are using `Linux`

Ay1nDas committed:
Removed trailing space from test21215.d

Ay1nDas committed:
Changed argLabels from _d_dynamicArray to Array in C++ side, updated front-end to match it.

Ay1nDas committed:
Removed typedef of ArgumentLabels from arraytypes.h

Ay1nDas committed:
Removed include frontend.h from  expression.h and removed extern (c++) from ArgumentLabel

Ay1nDas commented:
I've split the changes in 2 parts in this I've added `argLabels` and used `argLabels.loc` where it was required for the bug fix. 
Now in next pr I will replace `names` with `argLabels` .
I've decided to do this because this pr is focused towards fixing issue #21215

dkorpel commented:
Good, thanks for fixing the issue.
